tensorflow/contrib/tpu/python/tpu/keras_support.py

```diff
@@ -2145,20 +2145,23 @@ def load_weights(self, filepath, by_name=False):
 def _validate_shapes(model):
   """Validate that all layers in `model` have constant shape."""
   for layer in model.layers:
-    if isinstance(layer.input_shape, tuple):
-      input_shapes = [layer.input_shape]
-    else:
-      input_shapes = layer.input_shape
+    for node_idx in range(len(layer._inbound_nodes)):
```

Comment:
It seems like we want to accumulate the input and output shapes here. This appears to take only the last shape into account, and assumes the number of inputs and outputs are the same:

```
input_shapes = []
output_shapes =[]

for node_idx in range(...):
  input_shapes.append(layer.get_input_shape_at(node_idx))

for node_idx in range(...):
  output_shapes.append(layer.get_output_shape_at(node_idx)
```

For testing, I've found it is often easiest to use the TPU enable version of Colab, replace/sed the keras_support file and restart the notebook:

```
import tensorflow as tf
from tensorflow.contrib.tpu.python.tpu import keras_support
keras_support.__file__

with open('/usr/local/lib/python3.6/dist-packages/tensorflow/contrib/tpu/python/tpu/keras_support.py', 'w') as f:
  f.write(NEW_CONTENT)
```

Reply:
My understanding is that there are 2 concepts here, multiple nodes and multiple inputs/outputs, where multiple nodes means that the same layer (with the same weights) is being used in more than one location in the graph. I was assuming that everything was working correctly except for the fact that this code was assuming only one node per layer, so that's what this change was aimed at, but I will need to test it to confirm.

I ran into problems testing it on colab because I tried to just replace the whole file, but then it no longer compiles with tf 1.13rc2. It works with tf-nightly, but there appears to be a regression in tf-nightly that causes my model to no longer compile.

I'll keep working on it.